fast64_internal/util/collection_list_base.py

```diff
@@ -0,0 +1,247 @@
+) -> tuple[object, bpy.types.Collection, int]:
+    collection_base = context
+
+    for i in range(0, len(collection_name)):
+        if i > 0:
+            collection_base = col_prop[index]
+
+        collection_base = get_attr_from_dot_path(collection_base, base_path[i])
+        col_prop = get_attr_from_dot_path(collection_base, collection_name[i])
+        index = get_attr_from_dot_path(collection_base, index_name[i])
+
+    return collection_base, col_prop, index
+
+
+class CollectionOperator(bpy.types.Operator):
+    bl_options = {"REGISTER", "UNDO", "PRESET"}
+
+    base_path: str = None  # path from context to the root of your prop, e.g. context.[scene.base_prop]
+    collection_name: tuple[str] = None  # from base path, collection prop
+    index_name: tuple[str] = None  # prop that tracks index
+
+    def get_collection(self) -> tuple[object, bpy.types.Collection, int]:
+        if self.base_path is None:
+            raise UnimplementedError("self.base_path not defined")
+        if self.collection_name is None:
+            raise UnimplementedError("self.collection_name not defined")
+        if self.index_name is None:
+            raise UnimplementedError("self.index_name not defined")
+
+        return get_collection_props_from_context(bpy.context, self.base_path, self.collection_name, self.index_name)
+
+
+class AddNewToCollection(CollectionOperator):
+    bl_label = "Add an item to the list"
+
+    @staticmethod
+    def get_idname(base_list_name):
+        return f"{base_list_name}.add_item"
+
+    def execute(self, context):
+        _, collection, _ = self.get_collection()
+        collection.add()
+
+        return {"FINISHED"}
+
+
+class RemoveFromCollection(CollectionOperator):
+    bl_label = "Remove an item in the list"
+
+    @staticmethod
+    def get_idname(base_list_name):
+        return f"{base_list_name}.remove_item"
+
+    @classmethod
+    def poll(cls, context):
+        # TODO: Remove if __init__ is successful
+        raise Exception("classmethod `poll` not implemented: should poll the collection")
+
+    def execute(self, context):
+        collection_base, collection, index = self.get_collection()
+        collection.remove(index)
+        setattr(collection_base, self.index_name[-1], min(max(0, index - 1), len(collection)))
+
+        return {"FINISHED"}
+
+
+class MoveItemInCollection(CollectionOperator):
+    """Move an item in the list."""
+
+    bl_label = "Move an item in the list"
+
+    direction: bpy.props.EnumProperty(
+        items=(
+            ("UP", "Up", ""),
+            ("DOWN", "Down", ""),
+        )
+    )
+
+    @staticmethod
+    def get_idname(base_list_name):
+        return f"{base_list_name}.move_item"
+
+    def move_index(self):
+        """Move index of an item render queue while clamping it."""
+        collection_base, collection, index = self.get_collection()
+
+        list_length = len(collection) - 1  # (index starts at 0)
```

Comment:
what?
maybe this should be named `max_index` or `last_index` ?

Reply:
yeah makes much more sense. done